packages/react-client/src/hooks/useTrackManager.ts

```diff
@@ -33,25 +35,39 @@ export const useTrackManager = ({ mediaManager, tsClient }: TrackManagerConfig):
 
   const currentTrack = useMemo(() => getRemoteOrLocalTrack(tsClient, currentTrackId), [tsClient, currentTrackId]);
 
+  function clearMiddleware() {
+    clearMiddlewareFnRef.current?.();
+    clearMiddlewareFnRef.current = null;
+  }
+
+  function clearAndGetProcessedTrack(inputTrack: MediaStreamTrack): MediaStreamTrack {
+    clearMiddleware();
+    const { onClear, track } = currentTrackMiddleware?.(inputTrack) ?? {};
+    if (onClear) clearMiddlewareFnRef.current = onClear;
+    return track ?? inputTrack;
+  }
+
   async function setTrackMiddleware(middleware: TrackMiddleware): Promise<void> {
     const mediaTrack = mediaManager.getTracks()[0];
-
+    setCurrentTrackMiddleware(() => middleware);
     if (!currentTrack || !mediaTrack) return;
 
-    const trackToSet = middleware ? middleware(mediaTrack) : mediaTrack;
-    await tsClient.replaceTrack(currentTrack.trackId, trackToSet);
+    clearMiddleware();
+    const { onClear, track } = middleware?.(mediaTrack) ?? { track: mediaTrack };
+    if (onClear) clearMiddlewareFnRef.current = onClear;
 
-    setCurrentTrackMiddleware(() => middleware);
+    await tsClient.replaceTrack(currentTrack.trackId, track);
   }
 
   async function initialize(deviceId?: string) {
     await mediaManager?.start(deviceId ?? true);
     if (!currentTrackId) return;
     const newTrack = mediaManager.getTracks()[0];
-    await tsClient.replaceTrack(currentTrackId, newTrack);
+    await tsClient.replaceTrack(currentTrackId, clearAndGetProcessedTrack(newTrack));
```

Comment:
It replaces remote track, but what about local track? Would on preview screen?

Reply:
the preview one is processed in a different place, but i agree to unify